x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportDeleteJobAction.java

```diff
@@ -288,21 +293,32 @@ private void checkJobIsNotOpen(String jobId, ClusterState state) {
         }
     }
 
-    private void markJobAsDeletingIfNotUsed(String jobId, TaskId taskId, ActionListener<PutJobAction.Response> listener) {
+    private void deleteDatafeedIfNecessary(DeleteJobAction.Request deleteJobRequest, ActionListener<AcknowledgedResponse> listener) {
 
-        datafeedConfigProvider.findDatafeedIdsForJobIds(Collections.singletonList(jobId), ActionListener.wrap(
-                datafeedIds -> {
-                    if (datafeedIds.isEmpty() == false) {
-                        listener.onFailure(ExceptionsHelper.conflictStatusException("Cannot delete job [" + jobId + "] because datafeed ["
-                                + datafeedIds.iterator().next() + "] refers to it"));
-                        return;
-                    }
-                    cancelResetTaskIfExists(jobId, ActionListener.wrap(
-                        response -> jobConfigProvider.updateJobBlockReason(jobId, new Blocked(Blocked.Reason.DELETE, taskId), listener),
-                        listener::onFailure
-                    ));
-                },
-                listener::onFailure
+        datafeedConfigProvider.findDatafeedIdsForJobIds(Collections.singletonList(deleteJobRequest.getJobId()), ActionListener.wrap(
+            datafeedIds -> {
+                if (datafeedIds.isEmpty()) {
+                    listener.onResponse(AcknowledgedResponse.TRUE);
+                    return;
+                }
+                DeleteDatafeedAction.Request deleteDatafeedRequest = new DeleteDatafeedAction.Request(datafeedIds.iterator().next());
+                deleteDatafeedRequest.setForce(deleteJobRequest.isForce());
+                deleteDatafeedRequest.timeout(deleteJobRequest.timeout());
+                ClientHelper.executeAsyncWithOrigin(
+                    client,
+                    ClientHelper.ML_ORIGIN,
+                    DeleteDatafeedAction.INSTANCE,
+                    deleteDatafeedRequest,
+                    ActionListener.wrap(
+                        listener::onResponse,
+                        e -> listener.onFailure(ExceptionsHelper.conflictStatusException(
+                            "failed to delete job [{}] as its datafeed [{}] could not be deleted", e,
+                            deleteJobRequest.getJobId(), deleteDatafeedRequest.getDatafeedId())
+                        )
```

Comment:
Seems like we should handle "ResourceNotFound" here as the datafeed could have been deleted by a separate process. Would stink to fail the whole deletion due to a silly race condition where the datafeed is actually deleted.

Reply:
I am not sure the cases where we return `ResourceNotFound` it may be tricky if we return that if only ONE of the datafeeds are not found. 

Possibly the case `ResourceNotFound` && datafeedIds.size() == 1???

It seems to me that delete datafeed should simply take care of this already. But, looking through the code, there are cases where we will return `ResourceNotFound` when deleting a datafeed.

Reply:
You can only delete one job at a time at the moment, so there's at most one datafeed.  I'll add a comment about that.

Good point about `ResourceNotFound` - I'll add in a check for that.